File: 补充示例/scm_simulated.R
# Run from the companion root; data are simulated teaching examples.
library(quadprog)
df <- read.csv("data/processed/chapter13_scm_cities_simulated.csv")
Y <- xtabs(outcome ~ year + city, data=df)
pre <- as.integer(rownames(Y)) < 2016
A <- Y[pre,-1]; y <- Y[pre,1]; J <- ncol(A)
fit <- solve.QP(2*(crossprod(A)+1e-10*diag(J)),
  2*as.vector(crossprod(A,y)), cbind(rep(1,J),diag(J)),
  c(1,rep(0,J)), meq=1)
w <- fit$solution
stopifnot(min(w)>-1e-8,abs(sum(w)-1)<1e-8)
gap <- Y[,1]-Y[,-1] %*% w
print(w); print(mean(gap[pre]^2)); print(mean(gap[!pre]))
write.csv(data.frame(weight=w),"补充示例/scm_R.csv",row.names=FALSE)

File: 修订补充示例/scm_simulated.py
# Run from the companion root; data are simulated teaching examples.
import numpy as np, pandas as pd
from scipy.optimize import minimize
df = pd.read_csv("data/processed/chapter13_scm_cities_simulated.csv")
Y = df.pivot(index="year",columns="city",values="outcome").sort_index()
pre = Y.index < 2016
A = Y.loc[pre].iloc[:,1:].to_numpy(); y = Y.loc[pre,0].to_numpy()
J = A.shape[1]
res = minimize(lambda w: np.sum((y-A@w)**2)+1e-10*(w@w),
    np.ones(J)/J, jac=lambda w: 2*A.T@(A@w-y)+2e-10*w,
    bounds=[(0,1)]*J, constraints={"type":"eq","fun":lambda w: w.sum()-1},
    method="SLSQP", options={"ftol":1e-12,"maxiter":10000})
if not res.success: raise RuntimeError(res.message)
w = res.x
gap = Y[0].to_numpy()-Y.iloc[:,1:].to_numpy()@w
print(w, np.mean(gap[pre]**2), np.mean(gap[~pre]))
pd.DataFrame({"weight":w}).to_csv("修订补充示例/scm_Python.csv",index=False)
